### PR TITLE
Disable the GC after taking pool-related spinlocks.

### PR DESCRIPTION
**Pull request by maleadt**
Potential fix for https://github.com/JuliaGPU/CuArrays.jl/issues/685.

@denizyuret @colinxs @lpjiang97 could you try out this PR?

### CODE REVIEW

File: src/memory.jl [hunk 1]
```diff
@@ -11,6 +11,19 @@ using Base.Threads: SpinLock
 # each allocator needs to lock its own resources separately too.
 const memory_lock = SpinLock()
 
+# the above spinlocks are taken around code that might gc, which might cause a deadlock
+# if we try to acquire from the finalizer too. avoid that by temporarily disabling the GC.
+macro without_gc(ex)
+  quote
+    gc = GC.enable(false)
```
Comment (vchuravy) [inline src/memory.jl:18]: Julia base uses `jl_gc_enable_finalizers(ptls, 0)`.

```
ptls = Core.getptls()
ccall(:jl_gc_enable_finalizers, Cvoid, (Ptr{Cvoid}, Cint), ptls, 0)
```

Should be enough to disable finalizers to run. There is a question if the code needs to be sigatomic.
Looking at the code in locks.h I am a bit confused. It is unclear to me what happens when two threads try to acquire two different mutexes and they interleave the calls to `jl_gc_enable_finalizers`, similarly here we can't really guard against some-one enabling GC while we are in the critical section.

**Comment by denizyuret** [comment]
This solves my test case (it doesn't lock up any more).

File: src/memory.jl [hunk 2]
```diff
@@ -12,14 +12,15 @@ using Base.Threads: SpinLock
 const memory_lock = SpinLock()
 
 # the above spinlocks are taken around code that might gc, which might cause a deadlock
-# if we try to acquire from the finalizer too. avoid that by temporarily disabling the GC.
-macro without_gc(ex)
+# if we try to acquire from the finalizer too. avoid that by temporarily disabling finalizers.
```
Comment (vchuravy) [inline src/memory.jl:15]:
```suggestion
# if we try to acquire from the finalizer too. avoid that by temporarily disabling running finalizers,
# concurrently on this thread.
```

Chatting with Jameson about this, `jl_gc_enable_finalizer` will only inhibit running finalizers on the calling thread. They can still run on other threads (which should be fine). 

That is also the reason why the interleaving I was worried about can't happen.

**Comment by lpjiang97** [comment]
@maleadt Commit d0dc90 also solves the issue for me! But not the latest Commit 2d99e03. The deadlock still seems to happen. Here's the full stacktrace:

```julia

Stacktrace:
 [1] lock(::Base.Threads.SpinLock) at ./locks-mt.jl:71
 [2] macro expansion at ./lock.jl:181 [inlined]
 [3] free(::CUDAdrv.CuPtr{Nothing}) at /home/lpjiang97/.julia/packages/CuArrays/DO2vR/src/memory/binned.jl:393
 [4] macro expansion at /home/lpjiang97/.julia/packages/TimerOutputs/NvIUx/src/TimerOutput.jl:245 [inlined]
 [5] macro expansion at /home/lpjiang97/.julia/packages/CuArrays/DO2vR/src/memory.jl:245 [inlined]
 [6] macro expansion at ./util.jl:234 [inlined]
 [7] free at /home/lpjiang97/.julia/packages/CuArrays/DO2vR/src/memory.jl:244 [inlined]
 [8] _unsafe_free!(::CuArray{Float32,3,Nothing}) at /home/lpjiang97/.julia/packages/CuArrays/DO2vR/src/array.jl:51
 [9] unsafe_free!(::CuArray{Float32,3,Nothing}) at /home/lpjiang97/.julia/packages/CuArrays/DO2vR/src/array.jl:40
 [10] enable_finalizers at /home/lpjiang97/.julia/packages/CuArrays/DO2vR/src/memory.jl:16 [inlined]
 [11] macro expansion at /home/lpjiang97/.julia/packages/CuArrays/DO2vR/src/memory.jl:23 [inlined]
 [12] macro expansion at ./lock.jl:183 [inlined]
 [13] alloc(::Int64) at /home/lpjiang97/.julia/packages/CuArrays/DO2vR/src/memory/binned.jl:383
 [14] macro expansion at /home/lpjiang97/.julia/packages/TimerOutputs/NvIUx/src/TimerOutput.jl:245 [inlined]
 [15] macro expansion at /home/lpjiang97/.julia/packages/CuArrays/DO2vR/src/memory.jl:196 [inlined]
 [16] macro expansion at ./util.jl:234 [inlined]
 [17] alloc at /home/lpjiang97/.julia/packages/CuArrays/DO2vR/src/memory.jl:195 [inlined]
 [18] CuArray{Float32,2,P} where P(::UndefInitializer, ::Tuple{Int64,Int64}) at /home/lpjiang97/.julia/packages/CuArrays/DO2vR/src/array.jl:107
 [19] similar at /home/lpjiang97/.julia/packages/CuArrays/DO2vR/src/array.jl:139 [inlined]
 [20] similar at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/LinearAlgebra/src/adjtrans.jl:197 [inlined]
 [21] * at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/LinearAlgebra/src/matmul.jl:153 [inlined]
 [22] #1285 at /home/lpjiang97/.julia/packages/Zygote/4tJp5/src/lib/array.jl:292 [inlined]
 [23] (::Zygote.var"#3260#back#1287"{Zygote.var"#1285#1286"{CuArray{Float32,2,Nothing},CuArray{Float32,2,Nothing}}})(::CuArray{Float32,2,Nothing}) at /home/lpjiang97/.julia/packages/ZygoteRules/6nssF/src/adjoint.jl:49
 [24] Dense at /home/lpjiang97/.julia/packages/Flux/Fj3bt/src/layers/basic.jl:122 [inlined]
 [25] (::typeof(∂(invoke)))(::CuArray{Float32,2,Nothing}) at /home/lpjiang97/.julia/packages/Zygote/4tJp5/src/compiler/interface2.jl:0
 [26] Dense at /home/lpjiang97/.julia/packages/Flux/Fj3bt/src/layers/basic.jl:133 [inlined]
 [27] (::typeof(∂(λ)))(::CuArray{Float32,2,Nothing}) at /home/lpjiang97/.julia/packages/Zygote/4tJp5/src/compiler/interface2.jl:0
 [28] #3 at /home/lpjiang97/Research/probabilistic-pc/foo/debug.jl:21 [inlined]
 [29] (::typeof(∂(λ)))(::Float32) at /home/lpjiang97/.julia/packages/Zygote/4tJp5/src/compiler/interface2.jl:0
 [30] (::Zygote.var"#48#49"{Zygote.Params,Zygote.Context,typeof(∂(λ))})(::Float32) at /home/lpjiang97/.julia/packages/Zygote/4tJp5/src/compiler/interface.jl:108
 [31] gradient(::Function, ::Zygote.Params) at /home/lpjiang97/.julia/packages/Zygote/4tJp5/src/compiler/interface.jl:45
 [32] ISTA(::CuArray{Float32,2,Nothing}, ::Dense{typeof(identity),CuArray{Float32,2,Nothing},CuArray{Float32,1,Nothing}}, ::Float64, ::Int64, ::Float64, ::Float64) at /home/lpjiang97/Research/probabilistic-pc/foo/debug.jl:20
 [33] top-level scope at /home/lpjiang97/Research/probabilistic-pc/foo/debug.jl:48
 [34] include(::String) at ./client.jl:439
 [35] top-level scope at none:1
 [36] eval(::Module, ::Any) at ./boot.jl:331
 [37] exec_options(::Base.JLOptions) at ./client.jl:264
 [38] _start() at ./client.jl:484
```

**Comment by maleadt** [comment]
Alright, another attempt. Could you test again?

**Comment by lpjiang97** [comment]
This seems to have fixed it! Thanks @maleadt !

**Comment by denizyuret** [comment]
Still passes my tests.

**Pull request merged**